### PR TITLE
Automaterials: Partial fix for constructing floor over existing floor

### PR DESCRIPTION
**Pull request by DoctorVanGogh**
Partial fix for issue #721.

### CODE REVIEW

**Comment by PeridexisErrant** [comment]
./plugins/automaterial.cpp: Contains trailing whitespace: line 472

That's a real build failure; the others are docs being broken because they require pull #713.

**Comment by DoctorVanGogh** [comment]
My bad - you really need to fix the broken builds - I'm just ignoring the checks right now....

**Comment by DoctorVanGogh** [comment]
But I have to add.... complaining about a blank space at the end of a comment line is bordering on... obsessive ;)

**Comment by PeridexisErrant** [comment]
> My bad - you really need to fix the broken builds - I'm just ignoring the checks right now....

I agree!  But it's breaking because my huge pull hasn't bee merged yet, so I'm not pushing too hard ;)

> But I have to add.... complaining about a blank space at the end of a comment line is bordering on... obsessive ;)

Maybe, but with hundreds of people contributing if you're not anal about the little things it's way too easy to end up with a real mess.  Basically we try to break the build on literally any little issue we can detect, and hope that makes people more careful with the other stuff too ;)

**Comment by lethosor** [comment]
Also, some editors remove trailing whitespace automatically, so keeping it from being introduced prevents clutter when it gets removed later.

File: plugins/automaterial.cpp [hunk 1]
```diff
@@ -460,19 +461,27 @@ static bool is_valid_building_site(building_site &site, bool orthogonal_check, b
 
             if (material == tiletype_material::CONSTRUCTION)
             {
+                // Check this is not a  'constructed floor over X':
+                // those have something else than open space as *basic* tile types: stonefloor, smoothstonefloor, ...
+                MapExtras::MapCache mc;
+                auto btt = mc.baseTiletypeAt(site.pos);
+                auto bshape = tileShape(btt);
+                if (bshape != tiletype_shape_basic::Open)
```
Comment (lethosor) [inline plugins/automaterial.cpp]: From GCC: 
```
warning: comparison between 'enum df::enums::tiletype_shape::tiletype_shape' and 'enum df::enums::tiletype_shape_basic::tiletype_shape_basic'
```

Comment (DoctorVanGogh) [inline plugins/automaterial.cpp]: There's my general ignorance of df structures & types showing ;)

**Comment by DoctorVanGogh** [comment]
Fixes #721. Fixes #734.

No more floor over floor, and you can only construct *once* on top of a wall. Oh, and you can now build on top of down stairs.


Okay - this could probably be improved performance wise, but the basic functionality is in there. Theoretical worst case right now is iterating through all world construction for each and every designated tile. 
Some Caching? Doing this before checking individual tiles for validity en block like "Get all constructions inside these n candidate positions"?.

**Comment by lethosor** [comment]
I haven't gotten around to testing this yet - if you have time, does this fix the issue in #613?

**Comment by lethosor** [comment]
Looks good!

**Pull request merged**